### ISSUE TITLE
ffmpeg captures are choppy/laggy, streamlink as alternative? 

### ISSUE THREAD

**Issue by kuchenstrudel** [jrudess/streamdvr#27]
Using ffmpeg, many mfc recordings seems broken, meaning: choppy/laggy. 
With [streamlink](https://github.com/streamlink/streamlink) this issue is gone for me.

Example change to getCaptureArguments (site.js) might look like this.
> getCaptureArguments(url, filename) {
        return [
            "-l",
            "error",
            "hlsvariant://"+url,
            "best",
            "-o",
            this.config.captureDirectory + "/" + filename + ".ts"
        ];
    }
and of course changing `ffmpeg` to `streamlink` in the  `captureProcess` function.

This results in a perfect capture, but the recording stops after ~8 minutes (this was a quick and dirty hack, so maybe I messed something up).

Am I'm the only one with laggy/choppy captures?
Maybe there is something to be done with the ffmpeg spawn arguments?! 
Maybe a switch to streamlink would be a good idea?

**Comment by kuchenstrudel** [comment]
Side note: manually invoking `streamlink` wihout `streamdvr` doesn't break the capture after ~8 minutes.

**Comment by jrudess** [comment]
Flash is on its way out, so I don't think adding streamlink support is a very long term solution.   

I don't  currently have any issue with ffmpeg captures, so I would guess it is a difference with connections or route-to-mfc-server related.  But I've heard lots of other people mention that they get choppy recordings with ffmpeg and prefer rtmp for that reason.   

If you don't already have the latest version, try getting ffmpeg 3.4.2 and see if there is any change.

**Comment by Asynchr0nous** [comment]
ffmpeg has ALWAYS been an issue for recording MFC models so its not just you. for that reason i use streamlink although you do sometimes end up with choppy/laggy videos.

**Comment by kuchenstrudel** [comment]
I'm already using `ffmpeg 3.4.2`.  
`streamlink` also uses the mobile feed (hls) in my example. 
Bandwidth is also not an issue for me, routing also seems fine. 

I literally used the same url string as input for ffmpeg and streamlink and compared them. 
The former is choppy/laggy, the later is clear and stable.

**Comment by Asynchr0nous** [comment]
> Flash is on its way out, so I don't think adding streamlink support is a very long term solution.

@jrudess what do you mean by this?

**Comment by jrudess** [comment]
The write up on the streamlink dvr implies that it is for flash video.

`Streamlink is a CLI utility that pipes flash videos from online streaming services to a variety of video players such as VLC, or alternatively, a browser.`

If it also supports MFCs mobile streams, then that definitely is something different that I can look into.

**Comment by kuchenstrudel** [comment]
Streamlink supports _everything_. If you use `hlsvariant:// `as a prefix it works like a charm.

**Comment by jrudess** [comment]
I'm probably doing something wrong but not having any luck taking an HLS m3u8 link for mfc, prefixing it with hlsvariant:// and passing that to streamlink.  It just exits immediately with an error.

**Comment by jrudess** [comment]
Added streamlink support in master branch.   Set 'streamlink: true' in config.yml, make sure streamlink is in your path, and it should work for all sites, but needs some testing.

**Comment by jrudess** [comment]
I probably need to expose the m3u8 'sources' as a per-site config option, or possibly even per-model.  Currently I'm defaulting streamlink to use 'best' source.  

In the old mfc_cb.node repo I had opened an issue for having better control ofm3u8 sources.  It looks like streamlink is going to make implementing that very easy.

**Comment by kuchenstrudel** [comment]
Thanks! Will test and report back.

**Comment by kuchenstrudel** [comment]
First issue: if a stream goes offline "$model stream is off.", the streamlink process doesn't end. 
Second issue: if a capture for some reasons stops, the streamlink process doesn't end or restarts.

If you manually invoke the capturing process (without streamdvr) this issue seems to be gone.
Will do some more testing.

**Comment by jrudess** [comment]
When streamlink captured .ts are  converted to mp4 there is some kind of time-stamp problem.  The total time reported by video players is almost always wrong.  One file reports 13:55, but plays for 16:21.

I was able to find one streamlink recording that was 1.3GB however and was over an hour long, so on the surface it doesn't appear that I'm running into the same issue you are seeing yet.

I have not checked if streamlink responds to SIGINT requests, and that could explain processes hanging, since thats the method I use to kill ffmpeg processes that get hung up.

**Comment by jrudess** [comment]
I'm also noticing that a lot of files are missing audio in recordings, but am not sure if that is specific to streamlink yet.

**Comment by jrudess** [comment]
streamlink doesn't appear to respond to SIGINT requests, so an new method will need to be devised to terminate streamlink captures cleanly.

**Comment by kuchenstrudel** [comment]
Did you try SIGTERM? It does seem to work form the command line at least.
If this doesn't work, SIGKILL will, but it's not very clean.

**Comment by jrudess** [comment]
@kuchenstrudel seeing the same thing you are where streamlink doesn't quit but nothing continues recording.

**Comment by kuchenstrudel** [comment]
Same. It's very odd... recording with streamlink (w/o the help of streamdvr) works though. Very confusing.

**Comment by lukiluki123** [comment]
The normal (single thread) downloading of hls segments is quite slow with streamlink and after a while the segments die before streamlink would be able to download them. If you use option `--hls-segment-threads 5` it will work. 
Now I can record streams for hours, but without this option I was able to record for max 50-60 seconds

**Comment by jrudess** [comment]
Thanks for the info, added that switched and will see how it goes on this end.

**Comment by jrudess** [comment]
@kuchenstrudel Hoping you can try this too for an extra data-point.  

Unfortunately, the recordings still lock up for me with threads set to both 5 and 10 (kill -9 is still the only way to end the streamdvr process).

**Comment by kuchenstrudel** [comment]
Will test and report back.

**Comment by jrudess** [comment]
streamlink 0.13 is out with some fixes for ctrl+c and some changes to --hls-duration.   Going to try setting --hls-duration to a value shorter than the time I normally see the recording lock up, and hope there is an effect.

**Comment by kuchenstrudel** [comment]
`--hls-segment-threads 5` gives me at max around 22min of capture.
Will test `10` next. Also looking into all the other `--hls-*` options...

**Comment by kuchenstrudel** [comment]
tl;dr: I got it working.

I'm using the [hlssession](https://github.com/back-to/plugins/blob/master/plugins/hlssession.py) plugin to keep the session from hanging up, I think this was/is the issue...
`git clone https://github.com/back-to/plugins.git`
`cd plugins`
`ln -s "$(pwd)/plugins/hlssession.py" "$HOME/.config/streamlink/"`

This plugin needs to import `streamlink.utils.times` which is only in `streamlink` >=0.14.* (currently only available from github), so you need to get that too.

Also I replaced `SIGINT` with `SIGKILL`.

`hlsvariant://` needs to be replaced with `hlssession://`.
Parameters I'm using:
`"--hlssession-time",` (Reloads a Streamlink session after the given time.)
`"00:05:00",`
`"--hlssession-segment"` (Reloads a Streamlink session, if the playlist reload fails twice.)

I also added a way to `log` to see what `streamlink` does, makes debugging a bit easier. 

`diff --git a/core/site.js b/core/site.js`
```diff --git a/core/site.js b/core/site.js
index 910817e..5e1a9be 100644
--- a/core/site.js
+++ b/core/site.js
@@ -84,7 +84,7 @@ class Site {
             this.dbgMsg(colors.name(streamers.nm) + " file size (" + streamers.filename + "), size=" + stat.size + ", maxByteSize=" + maxByteSize);
             if (stat.size >= maxByteSize) {
                 this.msg(colors.name(streamers.nm) + " recording has exceeded file size limit (size=" + stat.size + " > maxByteSize=" + maxByteSize + ")");
-                streamers.captureProcess.kill("SIGINT");
+                streamers.captureProcess.kill("SIGKILL");
             }
         }
     }
@@ -97,13 +97,17 @@ class Site {
         let params = [];
 
         if (this.config.streamlink) {
-            const urlProt = "hlsvariant://" + url;
+            const urlProt = "hlssession://" + url;
             params = [
-                "-Q",
+                "-l",
+               "debug",
                 "-o",
                 this.config.captureDirectory + "/" + filename + ".ts",
                 urlProt,
-                "best"
+                "best",
+               "--hlssession-time",
+               "00:05:00",
+               "--hlssession-segment"
             ];
         } else {
             params = [
@@ -293,7 +297,7 @@ class Site {
     haltAllCaptures() {
         this.streamerList.forEach((value) => {
             if (value.captureProcess !== null) {
-                value.captureProcess.kill("SIGINT");
+                value.captureProcess.kill("SIGKILL");
             }
         });
     }
@@ -302,7 +306,7 @@ class Site {
         if (this.streamerList.has(uid)) {
             const streamer = this.streamerList.get(uid);
             if (streamer.captureProcess !== null) {
-                streamer.captureProcess.kill("SIGINT");
+                streamer.captureProcess.kill("SIGKILL");
             }
         }
     }
@@ -342,7 +346,10 @@ class Site {
     startCapture(streamer, filename, spawnArgs) {
         const fullname = filename + ".ts";
         const capper = this.config.streamlink ? "streamlink" : "ffmpeg";
+       var logStream = fs.createWriteStream("./" + filename + ".log", {flags: 'a'});
         const captureProcess = childProcess.spawn(capper, spawnArgs);
+       captureProcess.stdout.pipe(logStream);
+       captureProcess.stderr.pipe(logStream);
 
         if (captureProcess.pid) {
             this.msg(colors.name(streamer.nm) + " recording started (" + filename + ".ts)");
```

**Comment by jrudess** [comment]
Nice! Testing this out now.

**Comment by jrudess** [comment]
Made some tweaks to your change set
* New config.yml option -- debugrecorder: true
* Left SIGINT in place.  So far ctrl+c seems to be ending streamlink really well with 0.14.2

**Comment by SearchAndDestroy** [comment]
I am getting an error on Windows:
I recloned and just added a single model to test.
node is v10.5.0

`[cli][debug] OS:         Windows 10
[cli][debug] Python:     3.5.2
[cli][debug] Streamlink: 0.14.2
[cli][debug] Requests(2.19.1), Socks(1.6.7), Websocket(0.48.0)
[cli][info] Found matching plugin hlssession for URL hlssession://https://edge23.stream.highwebmedia.com/live-hls/amlst:testmodel-sd-cb57d909759a51d879e571354198802e2045ccea3bba690c4649f284daf7a4b6_trns_h264/chunklist_w1192668953_b1896000.m3u8


[cli][debug] Plugin specific arguments:
[cli][debug]  --hlssession-segment=True (segment)
[cli][debug]  --hlssession-time=300 (time)
[plugin.hlssession][info] This is a custom plugin. For support visit https://github.com/back-to/plugins
[plugin.resolve][info] This is a custom plugin. For support visit https://github.com/back-to/plugins
[plugin.resolve][debug] User-Agent: Mozilla/5.0 (Windows NT 6.1; Win64; x64; rv:52.0) Gecko/20100101 Firefox/52.0
[plugin.resolve][info]   1. URL=https://edge23.stream.highwebmedia.com/live-hls/amlst:testmodel-sd-cb57d909759a51d879e571354198802e2045ccea3bba690c4649f284daf7a4b6_trns_h264/chunklist_w1192668953_b1896000.m3u8

error: Unable to open URL: https://edge23.stream.highwebmedia.com/live-hls/amlst:testmodel-sd-cb57d909759a51d879e571354198802e2045ccea3bba690c4649f284daf7a4b6_trns_h264/chunklist_w1192668953_b1896000.m3u8
 (Invalid return character or leading space in header: Referer)
`

**Comment by kuchenstrudel** [comment]
Is this CB/Twitch/Mixer? I only tested MFC.

**Comment by SearchAndDestroy** [comment]
It's CB. I have not tested on Mixer or Twitch.
